adapters/mqtt-vertx-base/src/main/java/org/eclipse/hono/adapter/mqtt/CommandHandler.java

```diff
@@ -0,0 +1,170 @@
+            items.put(TracingHelper.TAG_QOS.getKey(), subscription.getQos().toString());
+            commandContext.getCurrentSpan().log(items);
+        });
+    }
+
+    /**
+     * Stores the published message id along with command subscription and command context.
+     *
+     * @param msgId The id of the command (message) that has been published.
+     * @param commandSubscription The device's command subscription.
+     * @param commandContext The commandContext of the command sent.
+     *
+     */
+    public void addToWaitingForAcknowledgement(final Integer msgId,
+            final CommandSubscription commandSubscription,
+            final CommandContext commandContext) {
+        waitingForAcknowledgement.put(msgId, TriTuple.of(startTimer(msgId), commandSubscription, commandContext));
+    }
+
+    /**
+     * Removes the entry from the waitingForAcknowledgement map for the given msgId.
+     *
+     * @param msgId The id of the command (message) that has been published.
+     * @return The trituple object containing timer-id, Commandsubscription and CommandContext for the given msgId.
+     */
+    public TriTuple<Long, CommandSubscription, CommandContext> removeFromWaitingForAcknowledgement(
+            final Integer msgId) {
+        return waitingForAcknowledgement.remove(msgId);
+    }
+    /**
+     * Stores the command subscription along with the command consumer.
+     *
+     * @param subscription The device's command subscription.
+     * @param commandConsumer A client for consuming messages.
+     */
+    public final void addSubscription(final CommandSubscription subscription,
+            final MessageConsumer commandConsumer) {
+        subscriptions.put(subscription.getTopic(), TriTuple.of(subscription, commandConsumer, null));
+    }
+
+    /**
+     * Closes the command consumer and removes the subscription entry for the given topic.
+     *
+     * @param topic The topic string to unsubscribe.
+     **/
+    public final void removeSubscription(final String topic) {
+        Optional.ofNullable(subscriptions.remove(topic)).ifPresent(value -> {
+            closeCommandConsumer(value.one(), value.two());
+        });
+    }
+
+    /**
+     * Stores the command subscription along with the command consumer.
+     *
+     * @param subscription The device's command subscription.
+     * @param commandConsumer A client for consuming messages.
+     */
+    private void closeCommandConsumer(final CommandSubscription subscription,
+            final MessageConsumer commandConsumer) {
+        commandConsumer.close(cls -> {
+            if (cls.succeeded()) {
+                LOG.trace("Command consumer closed [tenant-it: {}, device-id :{}]", subscription.getTenant(),
+                        subscription.getDeviceId());
+            } else {
+                LOG.error("Error closing command consumer [tenant-it: {}, device-id :{}]", subscription.getTenant(),
```

Comment:
this should not be logged at ERROR level. In fact, we don't really care about what the peer thinks about closing the link. We consider it closed in any case. So, IMHO this should be DEBUG ...